### DISCUSSION TITLE
transform texture maps for new connectivity

### DISCUSSION THREAD

**Discussion by chaohuang** [libigl/libigl#1936] [Q&A]
If the UV coordinates are changed but the connectivity is fixed, a new texture map can be obtained by calling `igl::map_texture`. My question is: how to get a new texture map if the connectivity is also changed? Is there any existing function in libigl or other libraries that can handle this situation?

Thanks!

**Comment by alecjacobson** [comment]
When the connectivity is the same then there's an obvious pointwise
correspondence between the two meshes (i.e., a point in any triangle of the
input mesh can be represented via barycentric coordinates, then its
corresponding point on the output mesh uses barycentric coordinates in the
corresponding triangle). So, if the connectivity changes, then a
correspondence would need to be explicitly specified. This is a non-trivial
problem and not implemented in libigl.
It really depends where your two meshes are coming from. If your
connectivity is changing via many local operations (like edge collapses and
flips) then you could try something like what's done in Neural
Subdivision/Surface Multigrid vis Intrinsic Prolongation. If the meshes are
totally different, then you might have to do something like "*Elastic
Correspondence between Triangle Meshes".*
-Alec

**Reply by chaohuang** [reply]
> 

Hi Prof. Jacobson,

Thanks for the quick reply!

The 2 meshes I have are very similar: `mesh1` is the original mesh, then its connectivity is removed to obtain 
a point cloud. After that, a surface reconstruction algorithm such as ball-pivoting is applied on the point cloud to estimate the original connectivity and we obtain `mesh2`. Therefore, all vertices of the 2 meshes are the same, only the connectivity is changed. 

So for this case, do you have any advice on how to transfer the texture map from `mesh1` to `mesh2`?

Thanks again!

**Reply by alecjacobson** [reply]
I see. Even this case might be tricky. The naive thing would be to use
closest point projection to map mesh2 onto mesh1 (you could use
igl::point_mesh_squared_distance for that). If the surfaces are really
close then this might work OK... except near original seam edges. Since the
new mesh2 doesn't have seams, much less in the right places, any triangle
in mesh2 that crosses a seam edge in its projection on mesh1 will look
really bad in UV space. Ideally, could subdivide those bad triangles at
just the right place to "cut" along the seam edge. Even if the projection
is bijective (very much not guaranteed with closest point) then you'd have
to walk along the projected triangle (no longer flat) and insert vertices
and edges whenever it crosses a seam edge. I'm not aware of existing code
for this. It might be the kind of thing that Mirela Ben Chen's group has
implemented.

**Reply by chaohuang** [reply]
Great! Thanks for the advice and I will look into this.
